File: android/src/main/kotlin/com/example/npu_pose_detection/VideoProcessor.kt
package com.example.npu_pose_detection

import android.content.Context
import android.graphics.Bitmap
import android.media.MediaMetadataRetriever
import android.util.Log
import com.example.npu_pose_detection.models.PoseResult
import kotlinx.coroutines.*
import java.io.File

/**
 * Video processor for frame-by-frame pose analysis.
 */
class VideoProcessor(
    private val context: Context,
    private val detector: PoseDetectorInterface
) {
    companion object {
        private const val TAG = "VideoProcessor"
    }

    private var isCancelled = false
    private var job: Job? = null
    private var progressCallback: ((VideoAnalysisProgress) -> Unit)? = null

    /**
     * Analyze a video file for poses.
     *
     * @param videoPath Path to the video file
     * @param frameInterval Process every Nth frame (1 = all frames)
     * @param progress Progress callback
     * @return VideoAnalysisResult with all detected poses
     */
    suspend fun analyzeVideo(
        videoPath: String,
        frameInterval: Int = 1,
        progress: (VideoAnalysisProgress) -> Unit
    ): VideoAnalysisResult = withContext(Dispatchers.IO) {
        progressCallback = progress
        isCancelled = false

        val file = File(videoPath)
        if (!file.exists()) {
            throw VideoProcessorException("File not found: $videoPath")
        }

        val retriever = MediaMetadataRetriever()
        try {
            retriever.setDataSource(videoPath)

            // Get video properties
            val durationMs = retriever.extractMetadata(
                MediaMetadataRetriever.METADATA_KEY_DURATION
            )?.toLongOrNull() ?: 0L

            val durationSeconds = durationMs / 1000.0

            // Get raw video dimensions (before rotation)
            val rawWidth = retriever.extractMetadata(
                MediaMetadataRetriever.METADATA_KEY_VIDEO_WIDTH
            )?.toIntOrNull() ?: 0

            val rawHeight = retriever.extractMetadata(
                MediaMetadataRetriever.METADATA_KEY_VIDEO_HEIGHT
            )?.toIntOrNull() ?: 0

            // Get rotation metadata - critical for portrait videos!
            val rotation = retriever.extractMetadata(
                MediaMetadataRetriever.METADATA_KEY_VIDEO_ROTATION
            )?.toIntOrNull() ?: 0

            // Calculate logical dimensions (as displayed to user)
            val (width, height) = if (rotation == 90 || rotation == 270) {
                rawHeight to rawWidth  // Swap dimensions for portrait videos
            } else {
                rawWidth to rawHeight
            }

            Log.i(TAG, "Video: raw=${rawWidth}x${rawHeight}, rotation=$rotation, logical=${width}x${height}")

            val frameRateString = retriever.extractMetadata(
                MediaMetadataRetriever.METADATA_KEY_CAPTURE_FRAMERATE
            )
            val frameRate = frameRateString?.toDoubleOrNull() ?: 30.0

            val totalFrames = (durationSeconds * frameRate).toInt()

            val frameResults = mutableListOf<VideoFrameResult>()
            var analyzedCount = 0
            val startTime = System.currentTimeMillis()

            // Calculate frame timestamps to extract
            val frameTimestampsUs = mutableListOf<Long>()
            var frameIndex = 0
            while (frameIndex < totalFrames) {
                if (isCancelled) break

                if (frameIndex % frameInterval == 0) {
                    val timestampUs = ((frameIndex.toDouble() / frameRate) * 1_000_000).toLong()
                    frameTimestampsUs.add(timestampUs)
                }
                frameIndex++
            }

            // Process each frame
            for ((index, timestampUs) in frameTimestampsUs.withIndex()) {
                if (isCancelled) break

                val actualFrameIndex = index * frameInterval
                val timestampSeconds = timestampUs / 1_000_000.0

                try {
                    // Extract frame at timestamp
                    var bitmap = retriever.getFrameAtTime(
                        timestampUs,
                        MediaMetadataRetriever.OPTION_CLOSEST
                    )

                    if (bitmap != null) {
                        // Apply rotation if needed - getFrameAtTime() returns unrotated frames
                        if (rotation != 0) {
                            bitmap = rotateBitmap(bitmap, rotation)
                        }

                        // Process the frame
                        val poseResult = detector.detectPose(bitmapToByteArray(bitmap))
                        bitmap.recycle()

                        frameResults.add(
                            VideoFrameResult(
                                frameIndex = actualFrameIndex,
                                timestampSeconds = timestampSeconds,
                                result = poseResult
                            )
                        )
                    }
                } catch (e: Exception) {
                    Log.w(TAG, "Failed to process frame at ${timestampSeconds}s: ${e.message}")
                    // Continue with next frame
                }

                analyzedCount++

                // Report progress
                val elapsed = System.currentTimeMillis() - startTime
                val framesPerSecond = if (elapsed > 0) analyzedCount * 1000.0 / elapsed else 0.0
                val remainingFrames = frameTimestampsUs.size - index - 1
                val estimatedRemaining = if (framesPerSecond > 0) {
                    remainingFrames / framesPerSecond
                } else null

                withContext(Dispatchers.Main) {
                    progress(
                        VideoAnalysisProgress(
                            currentFrame = actualFrameIndex,
                            totalFrames = totalFrames,
                            currentTimeSeconds = timestampSeconds,
                            durationSeconds = durationSeconds,
                            estimatedRemainingSeconds = estimatedRemaining
                        )
                    )
                }
            }

            val totalTime = (System.currentTimeMillis() - startTime).toInt()

            VideoAnalysisResult(
                frames = frameResults,
                totalFrames = totalFrames,
                analyzedFrames = analyzedCount,
                durationSeconds = durationSeconds,
                frameRate = frameRate,
                width = width,  // Logical width (after rotation)
                height = height,  // Logical height (after rotation)
                totalAnalysisTimeMs = totalTime
            )
        } finally {
            retriever.release()
        }
    }

    /**
     * Rotate bitmap by specified degrees.
     * MediaMetadataRetriever.getFrameAtTime() returns frames without rotation applied.
     */
    private fun rotateBitmap(bitmap: Bitmap, rotation: Int): Bitmap {
        if (rotation == 0) return bitmap

        val matrix = android.graphics.Matrix()
        matrix.postRotate(rotation.toFloat())
        val rotated = Bitmap.createBitmap(bitmap, 0, 0, bitmap.width, bitmap.height, matrix, true)
        bitmap.recycle()
        return rotated
    }

    /**
     * Cancel ongoing analysis.
     */
    fun cancel() {
        isCancelled = true
        job?.cancel()
    }

    private fun bitmapToByteArray(bitmap: Bitmap): ByteArray {
        val stream = java.io.ByteArrayOutputStream()
        bitmap.compress(Bitmap.CompressFormat.JPEG, 90, stream)
        return stream.toByteArray()
    }
}

/**
 * Exception thrown during video processing.
 */
class VideoProcessorException(message: String) : Exception(message)

/**
 * Progress during video analysis.
 */
data class VideoAnalysisProgress(
    val currentFrame: Int,
    val totalFrames: Int,
    val currentTimeSeconds: Double,
    val durationSeconds: Double,
    val estimatedRemainingSeconds: Double?
) {
    fun toMap(): Map<String, Any?> = mapOf(
        "type" to "progress",
        "currentFrame" to currentFrame,
        "totalFrames" to totalFrames,
        "currentTimeSeconds" to currentTimeSeconds,
        "durationSeconds" to durationSeconds,
        "estimatedRemainingSeconds" to estimatedRemainingSeconds
    )
}

/**
 * Result for a single video frame.
 */
data class VideoFrameResult(
    val frameIndex: Int,
    val timestampSeconds: Double,
    val result: PoseResult
) {
    fun toMap(): Map<String, Any> = mapOf(
        "frameIndex" to frameIndex,
        "timestampSeconds" to timestampSeconds,
        "result" to result.toMap()
    )
}

/**
 * Result of video analysis.
 */
data class VideoAnalysisResult(
    val frames: List<VideoFrameResult>,
    val totalFrames: Int,
    val analyzedFrames: Int,
    val durationSeconds: Double,
    val frameRate: Double,
    val width: Int,
    val height: Int,
    val totalAnalysisTimeMs: Int
) {
    fun toMap(): Map<String, Any> = mapOf(
        "frames" to frames.map { it.toMap() },
        "totalFrames" to totalFrames,
        "analyzedFrames" to analyzedFrames,
        "durationSeconds" to durationSeconds,
        "frameRate" to frameRate,
        "width" to width,
        "height" to height,
        "totalAnalysisTimeMs" to totalAnalysisTimeMs
    )
}

File: ios/Classes/VideoProcessor.swift
import Foundation
import AVFoundation

/// Video processor for frame-by-frame pose analysis.
class VideoProcessor {

    private var poseDetector: PoseDetectorProtocol?
    private var isCancelled = false
    private var progressHandler: ((VideoAnalysisProgress) -> Void)?
    private let processingQueue = DispatchQueue(label: "com.example.npu_pose_detection.video", qos: .userInitiated)

    init(detector: PoseDetectorProtocol) {
        self.poseDetector = detector
    }

    /// Analyze a video file for poses.
    ///
    /// - Parameters:
    ///   - videoPath: Path to the video file
    ///   - frameInterval: Process every Nth frame (1 = all frames)
    ///   - progress: Progress callback
    ///   - completion: Completion callback with result or error
    func analyzeVideo(
        at videoPath: String,
        frameInterval: Int = 1,
        progress: @escaping (VideoAnalysisProgress) -> Void,
        completion: @escaping (Result<VideoAnalysisResult, Error>) -> Void
    ) {
        self.progressHandler = progress
        self.isCancelled = false

        processingQueue.async { [weak self] in
            guard let self = self else { return }

            do {
                let result = try self.processVideo(at: videoPath, frameInterval: frameInterval)
                DispatchQueue.main.async {
                    completion(.success(result))
                }
            } catch {
                DispatchQueue.main.async {
                    completion(.failure(error))
                }
            }
        }
    }

    /// Cancel ongoing analysis.
    func cancel() {
        isCancelled = true
    }

    private func processVideo(at path: String, frameInterval: Int) throws -> VideoAnalysisResult {
        let url = URL(fileURLWithPath: path)

        guard FileManager.default.fileExists(atPath: path) else {
            throw VideoProcessorError.fileNotFound
        }

        let asset = AVAsset(url: url)

        // Get video properties
        guard let videoTrack = asset.tracks(withMediaType: .video).first else {
            throw VideoProcessorError.noVideoTrack
        }

        let duration = CMTimeGetSeconds(asset.duration)
        let frameRate = videoTrack.nominalFrameRate
        let totalFrames = Int(duration * Double(frameRate))
        let naturalSize = videoTrack.naturalSize

        // Create asset reader
        let reader = try AVAssetReader(asset: asset)

        let outputSettings: [String: Any] = [
            kCVPixelBufferPixelFormatTypeKey as String: kCVPixelFormatType_32BGRA
        ]

        let trackOutput = AVAssetReaderTrackOutput(track: videoTrack, outputSettings: outputSettings)
        trackOutput.alwaysCopiesSampleData = false
        reader.add(trackOutput)

        guard reader.startReading() else {
            throw VideoProcessorError.readerFailed(reader.error?.localizedDescription ?? "Unknown error")
        }

        var frameResults: [VideoFrameResult] = []
        var frameIndex = 0
        var analyzedCount = 0
        let startTime = CFAbsoluteTimeGetCurrent()

        while let sampleBuffer = trackOutput.copyNextSampleBuffer() {
            if isCancelled {
                reader.cancelReading()
                break
            }

            // Only process at specified interval
            if frameIndex % frameInterval == 0 {
                let timestamp = CMTimeGetSeconds(CMSampleBufferGetPresentationTimeStamp(sampleBuffer))

                do {
                    // Extract CVPixelBuffer from sample buffer
                    guard let pixelBuffer = CMSampleBufferGetImageBuffer(sampleBuffer) else {
                        throw VideoProcessorError.readerFailed("Failed to get pixel buffer")
                    }

                    let poseResult = try poseDetector?.processFrame(pixelBuffer: pixelBuffer)

                    if let result = poseResult {
                        let frameResult = VideoFrameResult(
                            frameIndex: frameIndex,
                            timestampSeconds: timestamp,
                            result: result
                        )
                        frameResults.append(frameResult)
                    }
                } catch {
                    // Continue on error, just skip the frame
                }

                analyzedCount += 1

                // Report progress
                let elapsed = CFAbsoluteTimeGetCurrent() - startTime
                let framesPerSecond = elapsed > 0 ? Double(analyzedCount) / elapsed : 0
                let remainingFrames = Double(totalFrames - frameIndex) / Double(frameInterval)
                let estimatedRemaining = framesPerSecond > 0 ? remainingFrames / framesPerSecond : nil

                let progress = VideoAnalysisProgress(
                    currentFrame: frameIndex,
                    totalFrames: totalFrames,
                    currentTimeSeconds: timestamp,
                    durationSeconds: duration,
                    estimatedRemainingSeconds: estimatedRemaining
                )

                DispatchQueue.main.async {
                    self.progressHandler?(progress)
                }
            }

            frameIndex += 1
        }

        let totalTime = Int((CFAbsoluteTimeGetCurrent() - startTime) * 1000)

        return VideoAnalysisResult(
            frames: frameResults,
            totalFrames: totalFrames,
            analyzedFrames: analyzedCount,
            durationSeconds: duration,
            frameRate: Double(frameRate),
            width: Int(naturalSize.width),
            height: Int(naturalSize.height),
            totalAnalysisTimeMs: totalTime
        )
    }
}

// MARK: - Error Types

enum VideoProcessorError: Error {
    case fileNotFound
    case noVideoTrack
    case readerFailed(String)
}

// MARK: - Data Models

struct VideoAnalysisProgress {
    let currentFrame: Int
    let totalFrames: Int
    let currentTimeSeconds: Double
    let durationSeconds: Double
    let estimatedRemainingSeconds: Double?

    func toDictionary() -> [String: Any] {
        var dict: [String: Any] = [
            "type": "progress",
            "currentFrame": currentFrame,
            "totalFrames": totalFrames,
            "currentTimeSeconds": currentTimeSeconds,
            "durationSeconds": durationSeconds
        ]
        if let remaining = estimatedRemainingSeconds {
            dict["estimatedRemainingSeconds"] = remaining
        }
        return dict
    }
}

struct VideoFrameResult {
    let frameIndex: Int
    let timestampSeconds: Double
    let result: PoseResult

    func toDictionary() -> [String: Any] {
        return [
            "frameIndex": frameIndex,
            "timestampSeconds": timestampSeconds,
            "result": result.toDictionary()
        ]
    }
}

struct VideoAnalysisResult {
    let frames: [VideoFrameResult]
    let totalFrames: Int
    let analyzedFrames: Int
    let durationSeconds: Double
    let frameRate: Double
    let width: Int
    let height: Int
    let totalAnalysisTimeMs: Int

    func toDictionary() -> [String: Any] {
        return [
            "frames": frames.map { $0.toDictionary() },
            "totalFrames": totalFrames,
            "analyzedFrames": analyzedFrames,
            "durationSeconds": durationSeconds,
            "frameRate": frameRate,
            "width": width,
            "height": height,
            "totalAnalysisTimeMs": totalAnalysisTimeMs
        ]
    }
}
